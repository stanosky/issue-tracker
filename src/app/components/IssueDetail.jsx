import React from "react";
import { Link } from "react-router-dom";

export class IssueDetail extends React.PureComponent {
    constructor(props) {
        super(props);
        this.state = {
            group: props.issue.group,
            name: props.issue.name,
            desc: props.issue.desc,
        };
        this.setTaskGroup = this.setTaskGroup.bind(this);
        this.setTaskName = this.setTaskName.bind(this);
        this.setTaskDesc = this.setTaskDesc.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
    }

    setTaskGroup(e) {
        this.setState({ group: e.target.value });
    }

    setTaskName(e) {
        this.setState({ name: e.target.value });
    }

    setTaskDesc(e) {
        this.setState({ desc: e.target.value });
    }

    onSubmit() {
        this.props.onUpdateTask(this.state);
    }

    render() {
        const { groups, issue } = this.props;
        return (
            <div className="card p-3 col-6">
                <div>
                    <span className="mr-4">Title:</span>
                    <input
                        type="text"
                        value={this.state.name}
                        onChange={this.setTaskName}
                        className="form-control form-control-lg"
                    />
                </div>
                <div>
                    <span className="mr-4">Description:</span>
                    <textarea
                        onChange={this.setTaskDesc}
                        defaultValue={this.state.desc}
                        className="form-control form-control-lg"
                    />
                </div>

                <span className="mr-4">Status: </span>
                <select onChange={this.setTaskGroup} className="form-control">
                    {groups.map((group) => (
                        <option
                            key={group.id}
                            value={group.id}
                            defaultValue={issue.group === group.id}
                            disabled={group.transitions.includes(issue.group)}
                        >
                            {group.name}
                        </option>
                    ))}
                </select>

                <div>
                    <button onClick={this.onSubmit} className="btn btn-primary mt-2">
                        Save
                    </button>
                    <Link to="/dashboard">
                        <button className="btn mt-2">Cancel</button>
                    </Link>
                </div>
            </div>
        );
    }
}
