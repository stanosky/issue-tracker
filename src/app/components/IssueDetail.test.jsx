import React from "react";
import { shallow } from "enzyme";
import { IssueDetail } from "./IssueDetail";

describe("IssueDetail", () => {
  const commonProps = {
    issue: {
      group: "g1",
      name: "name",
      desc: "desc",
    },
    groups: [
      {
        name: "test 1",
        id: "g1",
        transitions: ["g2"],
      },
      {
        name: "test 2",
        id: "g2",
        transitions: [],
      },
    ],
    onUpdateIssue: () => {},
  };
  it("should render without errors", () => {
    const wrapper = shallow(<IssueDetail {...commonProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should call onUpdateIssue method when button clicked", () => {
    const onUpdateIssue = jest.fn();
    const props = { ...commonProps, onUpdateIssue };
    const wrapper = shallow(<IssueDetail {...props} />);
    wrapper.find(".btn-primary").simulate("click");
    expect(onUpdateIssue).toHaveBeenCalledWith({
      group: "g1",
      name: "name",
      desc: "desc",
    });
  });

  it("should call onUpdateIssue method with correct params when status has been changed", () => {
    const onUpdateIssue = jest.fn();
    const props = { ...commonProps, onUpdateIssue };
    const wrapper = shallow(<IssueDetail {...props} />);
    wrapper.find("select").simulate("change", { target: { value: "g2" } });
    wrapper.find(".btn-primary").simulate("click");
    expect(onUpdateIssue).toHaveBeenCalledWith({
      group: "g2",
      name: "name",
      desc: "desc",
    });
  });
});
