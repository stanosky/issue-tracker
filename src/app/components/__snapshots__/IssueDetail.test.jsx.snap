// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueDetail should render without errors 1`] = `
<div
  className="card p-3 col-6"
>
  <div>
    <span
      className="mr-4"
    >
      Title:
    </span>
    <input
      className="form-control form-control-lg"
      onChange={[Function]}
      type="text"
      value="name"
    />
  </div>
  <div>
    <span
      className="mr-4"
    >
      Description:
    </span>
    <textarea
      className="form-control form-control-lg"
      defaultValue="desc"
      onChange={[Function]}
    />
  </div>
  <span
    className="mr-4"
  >
    Status: 
  </span>
  <select
    className="form-control"
    onChange={[Function]}
  >
    <option
      defaultValue={true}
      disabled={false}
      key="g1"
      value="g1"
    >
      test 1
    </option>
    <option
      defaultValue={false}
      disabled={false}
      key="g2"
      value="g2"
    >
      test 2
    </option>
  </select>
  <div>
    <button
      className="btn btn-primary mt-2"
      onClick={[Function]}
    >
      Save
    </button>
    <Link
      replace={false}
      to="/dashboard"
    >
      <button
        className="btn mt-2"
      >
        Cancel
      </button>
    </Link>
  </div>
</div>
`;
